Source/Events.Store.MongoDB/Processing/Streams/StreamProcessorStateDiscriminatorConvention.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using MongoDB.Bson;
+using MongoDB.Bson.IO;
+using MongoDB.Bson.Serialization.Conventions;
+
+namespace Dolittle.Runtime.Events.Store.MongoDB.Processing.Streams
+{
+    /// <summary>
+    /// Represents our own custom <see cref="IDiscriminatorConvention"/> used to deal with our MongoDB representation of
+    /// <see cref="StreamProcessorState"/> and <see cref="Partitioned.StreamProcessorStateExtensions"/>.
+    /// </summary>
+    /// <remarks>
+    /// This class was based mostly on information I gathered from these 2 sources:
+    /// https://groups.google.com/forum/#!topic/mongodb-user/iOeEXbUYbo4
+    /// https://github.com/mongodb/mongo-csharp-driver/blob/6b73e381827f83af368a949a807076dad01607fc/MongoDB.DriverUnitTests/Samples/MagicDiscriminatorTests.cs#L53 .
+    /// </remarks>
+    class StreamProcessorStateDiscriminatorConvention : IDiscriminatorConvention
+    {
+        public string ElementName => "Partitioned";
+
+        /// <summary>
+        /// Gets the value to be put into the ElementName key when writing <see cref="AbstractStreamProcessorState"/>
+        /// into the collection.
+        /// </summary>
+        /// <param name="nominalType">The nominal type.</param>
+        /// <param name="actualType">The actual type.</param>
+        /// <returns>Boolean indicating whether the StreamProcessorState is partitioned.</returns>
+        public BsonValue GetDiscriminator(Type nominalType, Type actualType) =>
+            actualType == typeof(Partitioned.PartitionedStreamProcessorState);
+
+        /// <summary>
+        /// Gets the correct type when deserializing objects from  <see cref="AbstractStreamProcessorState"/> collection
+        /// depending on the given nominal type or "Partitioned" field.
+        /// </summary>
+        /// <param name="bsonReader">A <see cref="IBsonReader"/>.</param>
+        /// <param name="nominalType">The nominal type.</param>
+        /// <returns>The actual wanted type.</returns>
+        public Type GetActualType(IBsonReader bsonReader, Type nominalType)
+        {
+            if (nominalType == typeof(StreamProcessorState) || nominalType == typeof(Partitioned.PartitionedStreamProcessorState))
+            {
+                return nominalType;
+            }
```

Comment:
To be honest - I think these cases should also be exceptions. This means that someone has registered this discriminator for any of those types, which is not right.

Reply:
Ohh - just saw your reply above - that you had seen `nominalType` be those two other ones. I find that surprising. What did you do to get those types in here?

Reply:
Just went into the `GetActualType` method with the debugger and saw that the `nominalType` was already the correct types sometimes.

The `GetActualType` method in our case is probably called from [`BsonClassMapSerializer`](https://github.com/mongodb/mongo-csharp-driver/blob/9e567e23615c8bb5c7ac1489427c2d15b2124522/src/MongoDB.Bson/Serialization/Serializers/BsonClassMapSerializer.cs#L126) with the [`BsonDeserializationArgs`](https://github.com/mongodb/mongo-csharp-driver/blob/010e7ee46b085cdd3762894ece9e2d258b66ab0d/src/MongoDB.Bson/Serialization/BsonDeserializationArgs.cs) struct that just contains the `nominalType`. I couldn't find where the `BsonDeserializationArgs` get created as they're kinda used everywhere in the repo.

Reply:
Removing the `[BsonKnownTypes]` fixes this and makes the `nominalType` always just `AbstractStreamProcessorState`